### ISSUE TITLE
[CUDA] max_depth parameter is ignored

### ISSUE THREAD

**Issue by lorenzolucchese** [microsoft/LightGBM#6969]
## Description
When training a lightgbm model with `device="cuda"` the `max_depth` parameter appears to be ignored. When using instead `device="cpu"` or `device="gpu"` (i.e. OpenCL implementation) the `max_depth` constraint is enforced.

## Reproducible example
```python
import lightgbm as lgb
from sklearn.datasets import make_regression

# Generate synthetic regression data
X, y = make_regression(n_samples=1000, n_features=20, noise=0.1, random_state=42)

# Create and fit the LGBMRegressor with GPU support
model = lgb.LGBMRegressor(
    objective="regression",
    device="cuda",  # Use CUDA
    max_depth=10,
)
model.fit(X, y)

print(model.booster_.trees_to_dataframe().groupby("tree_index")["node_depth"].max())
```

```
tree_index
0      8
1      9
2      9
3      9
4      9
      ..
95    18
96    14
97    18
98    17
99    14
Name: node_depth, Length: 100, dtype: int64
```

## Environment info

LightGBM version or commit hash: 4.6.0

Command(s) you used to install LightGBM

```shell
pixi add lightgbm
```

**Comment by jameslamb** [comment]
Thank you for the excellent report, and sorry you're facing this. I will try to reproduce this soon.

**Comment by Jessee-Jiao** [comment]
> Thank you for the excellent report, and sorry you're facing this. I will try to reproduce this soon.

So sorry, Has this bug been fixed? thanks

**Comment by shiyu1994** [comment]
The depth check is currently misisng in cuda_single_gpu_tree_learner.py, will fix it soon.

**Comment by jameslamb** [comment]
Thanks @shiyu1994 , I've assigned this to you.

It would be great to have a test that `max_depth` is actually enforced, maybe near these related tests:

https://github.com/microsoft/LightGBM/blob/a0fde1b00d40eb664945aeaca8c9159ff872206b/tests/python_package_test/test_basic.py#L911

That would be generally useful for all builds (in addition to specifically helping with this one for the CUDA variant).

**Comment by Jessee-Jiao** [comment]
I'm sorry, I have another question related to CUDA. I'm developing a program on CentOS Linux 7.9. My CUDA setup consists of two GeForce RTX 4090 GPUs. However, when I switched LightGBM from CPU to CUDA, the training speed barely improved. Is this normal, or am I missing some configuration details?

**Comment by jameslamb** [comment]
> _the training speed barely improved_

Many many things could cause this. GPU acceleration isn't guaranteed to make training faster.

Some reasons you might not get a speedup:

* your dataset is so small (in rows or columns) that the overhead of using the GPU (like copying data) overwhelms the speedups from using it to evaluate splits
* you are using custom objective or metric functions written in Python + running on the host, and these are slow enough that they dominate that training time
* what you're calling "training speed" is actually including the time of other operations in your code like reading data from files, sampling hyperparameter configurations, making plots, etc.

If you'd like us to investigate further, please **open a new issue** (so we can keep this one focused on `max_depth`) and try to provide a minimal, reproducible example (https://stackoverflow.com/help/minimal-reproducible-example). And please try to answer some of the questions like those mentioned here: https://github.com/microsoft/LightGBM/issues/6932#issuecomment-2942481286

"my code is slower than I expected" is very very broad, if you want us to give you some advice and investigate for possible performance issues, you need to work with us to help narrow things down.

**Comment by Jessee-Jiao** [comment]
Thank you very much for your detailed response. I will open a new issue later.
